### PR TITLE
container cleanup

### LINKED ISSUE #145: Cleanup of verlet-list functors and traversals
Make new verlet-list functors cleaner:
* extract possible common parts (similar to cell functors)
* reduce code duplication
* split traversals for verlet and linked-cells into different classes (#146)

### LINKED ISSUE #146: Cleanup traversals and containers
**Is your feature request related to a problem? Please describe.**
Everything is too cluttered

**Describe the solution you'd like**
Directory structure:
* Each container in a different directory
* Traversals of containers in the specific directories of the containers

### PR DESCRIPTION
**Pull request by SteffenSeckler**
# Description

Cleanup of Containers.

## Resolved Issues # (issue)

- [x] fixes #145 
- [x] fixes #146 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] adds tests for new examples

### CODE REVIEW

File: src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  /**
+   * Computes pairs for the block used in processBaseCell()
+   * @param cellsPerDimension
+   */
+  void computeOffsets(std::array<unsigned long, 3> cellsPerDimension);
+
+  /**
+   * CellFunctor to be used for the traversal defining the interaction between two cells.
+   */
+  CellFunctor<typename ParticleCell::ParticleType, ParticleCell, PairwiseFunctor, useSoA, useNewton3> _cellFunctor;
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h]: In some traversals, you call this `_cellFunctor` and in some `_cellProcessor` -> consistency plz :)

File: src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ * After executing the base step on all cells all pairwise interactions for
+ * all cells are done.
+ *
+ * @tparam ParticleCell the type of cells
+ * @tparam PairwiseFunctor The functor that defines the interaction of two particles.
+ * @tparam useSoA
+ * @tparam useNewton3
+ */
+template <class ParticleCell, class PairwiseFunctor, bool useSoA, bool useNewton3>
+class C08LikeBaseCellProcessor {
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h]: that's an awful name  ;)

File: src/autopas/containers/linkedCells/traversals/C18Traversal.h [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+ * @tparam useNewton3
+ */
+template <class ParticleCell, class PairwiseFunctor, bool useSoA, bool useNewton3>
+class C18Traversal : public C18BasedTraversal<ParticleCell, PairwiseFunctor, useSoA, useNewton3>,
+                     public LinkedCellTraversalInterface<ParticleCell> {
+ public:
+  /**
+   * Constructor of the c18 traversal.
+   * @param dims The dimensions of the cellblock, i.e. the number of cells in x,
+   * y and z direction.
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C18Traversal.h]: is this linebreak necessary?

File: src/autopas/containers/linkedCells/traversals/C18Traversal.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  /**
+   * @copydoc LinkedCellTraversalInterface::traverseCellPairs()
+   */
+  void traverseCellPairs(std::vector<ParticleCell> &cells) override;
+
+  /**
+   * Computes all interactions between the base
+   * cell and adjacent cells with greater a ID.
+   * @param cells vector of all cells.
+   * @param x x of base cell
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/C18Traversal.h]: x index

Comment (FG-TUM) [reply src/autopas/containers/linkedCells/traversals/C18Traversal.h]: Also applies to next lines

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h [hunk 5]
```diff
@@ -80,6 +80,11 @@ class VerletClusterLists : public ParticleContainer<Particle, FullParticleCell<P
    */
   template <class ParticleFunctor, class Traversal>
   void iteratePairwiseAoS(ParticleFunctor* f, Traversal* traversal, bool useNewton3 = true) {
+    // @todo
+    if (useNewton3) {
+      AutoPasLog(error, "Newton3 not implemented yet");
+      autopas::utils::ExceptionHandler::exception("VerletClusterLists does not support newton3.");
```
Comment (FG-TUM) [inline src/autopas/containers/verletClusterLists/VerletClusterLists.h]:
```suggestion
      autopas::utils::ExceptionHandler::exception("VerletClusterLists does not support newton3 yet.");
```

File: examples/sph-traversals/CMakeLists.txt [hunk 6]
```diff
@@ -10,3 +10,89 @@ ADD_EXECUTABLE(sph-traversals
 
 TARGET_LINK_LIBRARIES(sph-traversals
         autopas)
+
+
+#-----------------test-----------------
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply examples/sph-traversals/CMakeLists.txt]: https://github.com/AutoPas/AutoPas/pull/148/commits/dc50d8d4752387eb8c21ea19d16c5a084eb95786

File: examples/cluster-test/CMakeLists.txt [hunk 7]
```diff
@@ -10,3 +10,14 @@ ADD_EXECUTABLE(cluster-test
 
 TARGET_LINK_LIBRARIES(cluster-test
         autopas)
+
+#-----------------test-----------------
+# add check for current target
+add_test(
+        NAME cluster-test.test
+        COMMAND cluster-test 2000 0.10 0.1
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply examples/cluster-test/CMakeLists.txt]: https://github.com/AutoPas/AutoPas/pull/148/commits/3b8a26fb850861ca463988645c5657c8d97960c7

File: src/autopas/containers/cellPairTraversals/SlicedBasedTraversal.h [hunk 8]
```diff
@@ -30,31 +29,32 @@ namespace autopas {
...
+  bool isApplicable() final { return this->_sliceThickness.size() > 0; }
+
+  void rebuild(const std::array<unsigned long, 3> &dims) override;
+
+ protected:
   /**
-   * @copydoc VerletListsCellsTraversal::traverseCellVerlet
+   * The main traversal of the C01Traversal. This provides the structure of the loops and its parallelization.
+   * @tparam LoopBody
+   * @param loopBody the loop body. Normally a lambda function, that takes as as parameters (x,y,z). If you need
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/containers/cellPairTraversals/SlicedBasedTraversal.h]: https://github.com/AutoPas/AutoPas/pull/148/commits/5e772a418d8b4bb580d9f4a31e73dfe2edeefee2

File: src/autopas/containers/linkedCells/LinkedCells.h [hunk 9]
```diff
@@ -105,8 +110,12 @@ class LinkedCells : public ParticleContainer<Particle, ParticleCell, SoAArraysTy
   void iteratePairwiseSoA(ParticleFunctor *f, Traversal *traversal, bool useNewton3 = true) {
     AutoPasLog(debug, "Using traversal {} with SoA ", traversal->getTraversalType());
     loadSoAs(f);
-
-    traversal->traverseCellPairs(this->_cells);
+    if (auto *traversalInterface = dynamic_cast<LinkedCellTraversalInterface<ParticleCell> *>(traversal)) {
+      traversalInterface->traverseCellPairs(this->_cells);
+    } else {
+      autopas::utils::ExceptionHandler::exception(
+          "trying to use a traversal of wrong type in LinkedCells::iteratePairwiseSoA");
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/LinkedCells.h]: https://github.com/AutoPas/AutoPas/pull/148/commits/061f72e3b580315a306fd995a73ef7c441ddc2ac

File: src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h (hunk 1 above)
Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h]: invalid.
_cellFunctor is an actual cell functor
_baseCellProcessor handles the processBaseCell, which is built upon traversals.

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h (hunk 5 above)
Comment (SteffenSeckler) [reply src/autopas/containers/verletClusterLists/VerletClusterLists.h]: https://github.com/AutoPas/AutoPas/pull/148/commits/061f72e3b580315a306fd995a73ef7c441ddc2ac

File: src/autopas/containers/linkedCells/traversals/C18Traversal.h (hunk 4 above)
Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/traversals/C18Traversal.h]: https://github.com/AutoPas/AutoPas/pull/148/commits/64410a25a456134d1c66ddd20243ffd22c1ecbc4

File: src/autopas/containers/linkedCells/traversals/C18Traversal.h (hunk 3 above)
Comment (SteffenSeckler) [reply src/autopas/containers/linkedCells/traversals/C18Traversal.h]: https://github.com/AutoPas/AutoPas/pull/148/commits/38e0bd45b346714ffa11f4b7a23c0c9a0b80ec8b

File: src/autopas/containers/cellPairTraversals/C18BasedTraversal.h [hunk 10]
```diff
@@ -34,98 +33,55 @@ class C18BasedTraversal : public CellPairTraversal<ParticleCell> {
...
+    for (unsigned long col = 0; col < 18; ++col) {
+      std::array<unsigned long, 3> start = utils::ThreeDimensionalMapping::oneToThreeD(col, stride);
+
+      // intel compiler demands following:
+      const unsigned long start_x = start[0], start_y = start[1], start_z = start[2];
+      const unsigned long end_x = end[0], end_y = end[1], end_z = end[2];
+      const unsigned long stride_x = stride[0], stride_y = stride[1], stride_z = stride[2];
+
+#if defined(AUTOPAS_OPENMP)
+      // @todo: find optimal chunksize
```
[the remark this answers is not in the document]
Comment (SteffenSeckler) [reply src/autopas/containers/cellPairTraversals/C18BasedTraversal.h]: #150

File: src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h (hunk 2 above)
Comment (FG-TUM) [reply src/autopas/containers/linkedCells/traversals/C08LikeBaseCellProcessor.h]: C08CellProcessor

**Pull request merged**